source/extensions/dynamic_modules/abi.h

```diff
@@ -345,6 +379,135 @@ envoy_dynamic_module_on_http_filter_response_trailers(
 void envoy_dynamic_module_on_http_filter_destroy(
     envoy_dynamic_module_type_http_filter_module_ptr filter_module_ptr);
 
+// -----------------------------------------------------------------------------
+// -------------------------------- Callbacks ----------------------------------
+// -----------------------------------------------------------------------------
+//
+// Callbacks are functions implemented by Envoy that can be called by the module to interact with
+// Envoy. The name of a callback must be prefixed with "envoy_dynamic_module_callback_".
+
+/**
+ * envoy_dynamic_module_callback_http_get_request_header is called by the module to get the
+ * value of the request header with the given key. Since a header can have multiple values, the
+ * index is used to get the specific value. This returns the number of values for the given key, so
+ * it can be used to iterate over all values by starting from 0 and incrementing the index until the
+ * return value.
+ *
+ * PRECONDITION: Envoy does not check the validity of the key as well as the result_buffer_ptr
+ * and result_buffer_length_ptr. The module must ensure that these values are valid, e.g.
+ * non-null pointers.
+ *
+ * @param filter_envoy_ptr is the pointer to the DynamicModuleHttpFilter object of the
+ * corresponding HTTP filter.
+ * @param key is the key of the request header.
+ * @param key_length is the length of the key.
+ * @param result_buffer_ptr is the pointer to the pointer variable where the pointer to the buffer
+ * of the value will be stored. If the key does not exist or the index is out of range, this will be
+ * set to nullptr.
+ * @param result_buffer_length_ptr is the pointer to the variable where the length of the buffer
+ * will be stored. If the key does not exist or the index is out of range, this will be set to 0.
+ * @param index is the index of the header value in the list of values for the given key.
+ * @return the number of values for the given key, regardless of whether the value is found or not.
+ *
+ * Note that a header value is not guaranteed to be a valid UTF-8 string. The module must be careful
+ * when interpreting the value as a string in the language of the module.
+ *
+ * The buffer pointed by the pointer stored in result_buffer_ptr is owned by Envoy, and they are
+ * guaranteed to be valid until the end of the current event hook.
+ */
+size_t envoy_dynamic_module_callback_http_get_request_header(
```

Comment:
note: the choice of not defining "status" from this function (and the other variants) is intentional. This getter function is more of an abstraction over Envoy::Http::HeaderMap::get method, and that only returns a vector without status. In other words, the only "failure" mode is the not found case, which is already covered by the current definition. If the necessity comes up later, we can switch to defining a dedicated enum status, but for now we should be fine with this (we can break the ABI anyways).

This somewhat complex signature is not visible and hidden from the users at the SDK level.